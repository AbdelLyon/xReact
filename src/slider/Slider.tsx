import { useState, useEffect } from "react";

import { Slider as SliderRoot } from "@/ui/slider";
import { cn } from "@/utils";

export interface SliderValues {
  min: number;
  max: number;
}

export interface SliderProps {
  className?: string;
  values: SliderValues;
  onChange?: (values: SliderValues) => void;
  max?: number;
  step?: number;
  min?: number;
  showTooltips?: boolean;
}

export function Slider({
  className,
  onChange,
  values = {
    min: 0,
    max: 0,
  },
  max = 100,
  min = 0,
  step = 1,
  showTooltips,
  ...props
}: SliderProps) {
  const [selectedValues, setSelectedValues] = useState<SliderValues>(values);

  useEffect(() => {
    setSelectedValues(values);
  }, [values]);

  const handleValueChange = (newValues: number[]) => {
    const updatedValues = {
      min: newValues[0],
      max: newValues[1],
    };

    setSelectedValues(updatedValues);
    onChange?.(updatedValues);
  };

  return (
    <SliderRoot
      className={cn("w-full mt-10 relative", className)}
      max={max}
      min={min}
      step={step}
      value={[selectedValues.min, selectedValues.max]}
      onValueChange={handleValueChange}
      {...props}
    >
      {showTooltips && (
        <>
          <div
            className="absolute -top-9 left-0 -translate-x-1/2"
            style={{
              left: `${((selectedValues.min - min) / (max - min)) * 100}%`,
            }}
          >
            <div className="rounded-sm bg-black/80 px-2 py-1 text-xs text-white after:absolute after:left-1/2 after:top-full after:-translate-x-1/2 after:border-4 after:border-x-transparent after:border-b-transparent after:border-t-black/80 after:content-['']">
              {selectedValues.min}
            </div>
          </div>
          <div
            className="absolute -top-9 right-0 translate-x-1/2"
            style={{
              right: `${100 - ((selectedValues.max - min) / (max - min)) * 100}%`,
            }}
          >
            <div className="rounded-sm bg-black/80 px-2 py-1 text-xs text-white after:absolute after:left-1/2 after:top-full after:-translate-x-1/2 after:border-4 after:border-x-transparent after:border-b-transparent after:border-t-black/80 after:content-['']">
              {selectedValues.max}
            </div>
          </div>
        </>
      )}
    </SliderRoot>
  );
}
