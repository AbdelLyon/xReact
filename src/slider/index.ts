import "@/index.css";
export { Slider, type SliderValues } from "@/slider/Slider";
