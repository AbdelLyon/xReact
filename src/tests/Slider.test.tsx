import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import { Slider } from "@/slider";

describe("Slider component", () => {
  it("matches the snapshot", () => {
    const { asFragment } = render(<Slider values={{ min: 20, max: 80 }} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("calls onChange with correct values when dragging", async () => {
    const mockOnChange = vi.fn();

    render(<Slider values={{ min: 20, max: 80 }} onChange={mockOnChange} />);

    const sliders = screen.getAllByRole("slider");
    const leftSlider = sliders[0];

    fireEvent.keyDown(leftSlider, { key: "ArrowRight" });
    fireEvent.keyDown(leftSlider, { key: "ArrowRight" });

    await waitFor(() => {
      expect(mockOnChange).toHaveBeenCalled();
    });

    const lastCall =
      mockOnChange.mock.calls[mockOnChange.mock.calls.length - 1][0];

    expect(lastCall.min).toBeGreaterThan(20);
    expect(lastCall.max).toBe(80);
  });

  it("updates value on keyboard interaction", async () => {
    const mockOnChange = vi.fn();

    render(<Slider values={{ min: 20, max: 80 }} onChange={mockOnChange} />);

    const sliders = screen.getAllByRole("slider");
    const minSlider = sliders[0];

    fireEvent.keyDown(minSlider, { key: "ArrowRight" });
    fireEvent.keyDown(minSlider, { key: "ArrowRight" });

    await waitFor(() => {
      expect(mockOnChange).toHaveBeenCalled();
    });

    const lastCall =
      mockOnChange.mock.calls[mockOnChange.mock.calls.length - 1][0];

    expect(lastCall.min).toBeGreaterThan(20);
    expect(lastCall.max).toBe(80);
  });
});
