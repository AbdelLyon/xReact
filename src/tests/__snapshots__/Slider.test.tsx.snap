// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Slider component > matches the snapshot 1`] = `
<DocumentFragment>
  <span
    aria-disabled="false"
    class="flex touch-none select-none items-center w-full mt-10 relative"
    data-orientation="horizontal"
    dir="ltr"
    style="--radix-slider-thumb-transform: translateX(-50%);"
  >
    <span
      class="relative h-2 w-full grow overflow-hidden rounded-full bg-card-light dark:bg-secondary"
      data-orientation="horizontal"
    >
      <span
        class="absolute h-full bg-primary"
        data-orientation="horizontal"
        style="left: 20%; right: 20%;"
      />
    </span>
    <span
      style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(20% + 0px);"
    >
      <span
        aria-label="Minimum"
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="20"
        class="block size-5 rounded-full border-2 border-primary bg-background ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50"
        data-orientation="horizontal"
        data-radix-collection-item=""
        role="slider"
        style=""
        tabindex="0"
      />
    </span>
    <span
      style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(80% + 0px);"
    >
      <span
        aria-label="Maximum"
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="80"
        class="block size-5 rounded-full border-2 border-primary bg-background ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50"
        data-orientation="horizontal"
        data-radix-collection-item=""
        role="slider"
        style=""
        tabindex="0"
      />
    </span>
  </span>
</DocumentFragment>
`;
